File: thousands-services/IvsIdleGameShared/Models/Market/CreditTransaction.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace IvsIdleGameShared.Models.Market
{
    [BsonIgnoreExtraElements]
    public class CreditTransaction
    {
        [BsonId]
        [BsonElement("_id")]
        public ObjectId Id { get; set; }

        [BsonElement("userId")]
        public ObjectId UserId { get; set; }

        [BsonElement("transactionId")]
        public string TransactionId { get; set; } = "";

        [BsonElement("amount")]
        public int Amount { get; set; } = 0;

        [BsonElement("currency")]
        public string Currency { get; set; } = "";

        [BsonElement("paymentMethod")]
        public string PaymentMethod { get; set; } = "";

        [BsonElement("paymentGateway")]
        public string PaymentGateway { get; set; } = "";

        [BsonElement("paymentGatewayTransactionId")]
        public string? PaymentGatewayTransactionId { get; set; } = null;

        [BsonElement("refundedAmount")]
        public int RefundedAmount { get; set; } = 0;

        [BsonElement("status")]
        public string Status { get; set; } = "";

        [BsonElement("creditType")]
        public string? CreditType { get; set; } = null;

        [BsonElement("stageId")]
        public string? StageId { get; set; } = null;

        [BsonElement("segment")]
        public int? Segment { get; set; } = null;

        [BsonElement("skyboxTier")]
        public int? SkyboxTier { get; set; } = null;


        [BsonElement("createdAt")]
        public DateTime CreatedAt { get; set; } = DateTime.MinValue;

        [BsonElement("updatedAt")]
        public DateTime UpdatedAt { get; set; } = DateTime.MinValue;

        [BsonElement("__v")]
        public int Version { get; set; } = 0;
    }

    public static class CreditTransactionType
    {
        public const string CREDIT = "credit";
        public const string BOOST = "boost";
        public const string SKYBOX_PURCHASE = "skybox_purchase";
        public const string SPONSORSHIP_PURCHASE = "sponsorship_purchase";
        public const string ADMIN_ADJUSTMENT = "admin_adjustment";
    }
}

File: thousands-tv/wildcard/packages/interfaces/src/db/interfaces/creditTransaction.ts
import { Types } from "mongoose";

// inline with Thirdweb BuyWithCryptoTransaction
export enum CreditTransactionStatus {
    COMPLETED = "COMPLETED",
    PENDING = "PENDING",
    FAILED = "FAILED",
    REFUNDED = "REFUNDED",
    NONE = "NONE",
}

export enum CreditTransactionType {
    CREDIT = "credit",
    BOOST = "boost",
    ADMIN_ADJUSTMENT = "admin_adjustment",
    SKYBOX_PURCHASE = "skybox_purchase",
}

export interface ICreditTransaction {
    userId: Types.ObjectId; // User's unique identifier
    transactionId: string; // Unique transaction identifier for the credits transaction (for mongo reference)
    status: CreditTransactionStatus; // Status of the credit transaction
    amount: number; // Amount of money paid (in original currency)
    currency: string; // Currency used for the payment (e.g., USD, ETH)
    paymentGatewayTransactionId?: string; // Unique transaction identifier for this credit addition (from the payment gateway)
    paymentMethod?: string; // Payment method used (e.g., 'credit card', 'crypto')
    paymentGateway?: string; // Payment gateway used (e.g., "Thirdweb", 'Stripe', 'Coinbase')
    refundedAmount?: number; // Amount refunded (if applicable)
    creditType: CreditTransactionType; // Type of credit transaction
    identityId?: Types.ObjectId; // Identity's unique identifier

    adjustmentReason?: string; // Required for admin adjustments
    adjustedBy?: Types.ObjectId; // Admin user who made the adjustment

    stageId?: string;
    segment?: number;
    skyboxTier?: number;

    _id?: Types.ObjectId;
    __v?: number;
    createdAt?: Date;
    updatedAt?: Date;
}
